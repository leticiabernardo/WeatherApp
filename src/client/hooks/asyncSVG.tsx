import { useRef, useState, useEffect } from 'react';

interface UseDynamicSVGImportOptions {
  onCompleted?: (
    name: string,
    SvgIcon: React.FC<React.SVGProps<SVGSVGElement>> | undefined
  ) => void;
  onError?: (err: Error) => void;
}

export function useDynamicSVGImport(
  name: string,
  options: UseDynamicSVGImportOptions = {}
) {
  const ImportedIconRef = useRef<React.FC<React.SVGProps<SVGSVGElement>>>();
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<Error>();

  const { onCompleted, onError } = options;
  useEffect(() => {
    setLoading(true);
    const importIcon = async (): Promise<void> => {
      try {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-member-access
        ImportedIconRef.current = (
          await import(`../../assets/images/${name}.svg`)
        ).ReactComponent;
        onCompleted?.(name, ImportedIconRef.current);
      } catch (e) {
        const err = e as Error;
        onError?.(err);
        setError(err);
      } finally {
        setLoading(false);
      }
    };
    importIcon();
  }, [name, onCompleted, onError]);

  return { error, loading, SvgIcon: ImportedIconRef.current };
}
