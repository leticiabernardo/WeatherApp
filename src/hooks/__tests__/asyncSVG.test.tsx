import { render, cleanup, waitFor } from '@testing-library/react';
import { useDynamicSVGImport } from '../asyncSVG';

function TestComponent({ name }: { name: string }) {
  const { error, loading, SvgIcon } = useDynamicSVGImport(name);
  if (loading) {
    return <div>loading</div>;
  }
  if (SvgIcon) return <SvgIcon />;

  return <div>{error?.message}</div>;
}

describe('hook async svg test', () => {
  afterEach(cleanup);

  test('should return a component', async () => {
    const { container } = await waitFor(() =>
      render(<TestComponent name="icon-search.svg" />)
    );

    expect(container).toMatchSnapshot();
  });

  test('should return error', async () => {
    const { container } = await waitFor(() =>
      render(<TestComponent name="icon-search.error" />)
    );

    expect(container).toMatchSnapshot();
  });
});
