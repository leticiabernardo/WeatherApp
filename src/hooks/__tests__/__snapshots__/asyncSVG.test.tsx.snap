// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hook async svg test should return a component 1`] = `
<div>
  <div
    class="svg-mock"
  />
</div>
`;

exports[`hook async svg test should return error 1`] = `
<div>
  <div>
    erro
  </div>
</div>
`;
